use exact match for illegal path check

zhangbo1882 opened the pull request:
In our environment, the file system directory is as follows:
```
Tue Jun 04 22:28:35][#48# ]$df -h
Filesystem                 Size  Used Avail Use% Mounted on
tmpfs                       77G  104K   77G   1% /dev/shm
tmpfs                       31G  9.8M   31G   1% /run
tmpfs                      5.0M     0  5.0M   0% /run/lock
tmpfs                      4.0M     0  4.0M   0% /sys/fs/cgroup
/dev/mapper/atomicos-root  150G  144G  5.8G  97% /sysroot
/dev/vda2                  483M   84M  400M  18% /boot
/dev/vdc                   1.2T   87G  1.1T   8% /sysroot/home/centos/external
```
We have a directory named /sysroot. If the envoy config file is the that directory, envoy can not start up.
```
[2024-06-04 22:28:35.581][3382724][critical][main] [source/server/server.cc:131] error initializing configuration 'configs/envoy.yaml': Invalid path: configs/envoy.yaml
[2024-06-04 22:28:35.581][3382724][info][main] [source/server/server.cc:972] exiting
Invalid path: configs/envoy.yaml
```
In my mind, envoy should only check the default system directory such as /dev /sys /proc as illegal path.
So it is better to use exact match instead of startwith match.

adisuissa commented on source/common/filesystem/posix/filesystem_impl.cc:
```diff
@@ -339,9 +339,8 @@ bool InstanceImplPosix::illegalPath(const std::string& path) {
-  if (absl::StartsWith(canonical_path.return_value_, "/dev") ||
-      absl::StartsWith(canonical_path.return_value_, "/sys") ||
-      absl::StartsWith(canonical_path.return_value_, "/proc")) {
+  if (canonical_path.return_value_ == "/dev" || canonical_path.return_value_ == "/sys" ||
```
I think this now allows the file to be at "/dev/abc/my_file" which is what this check intended to avoid in the first place.
The check could be for startwith("/sys/") or equals("/sys"). I believe this should align the intentions of #3165 and your need to support the "/sysroot" directory.
cc @htuch as the author of #3165.

zhangbo1882 replied:
Thanks @adisuissa , I will update it.

adisuissa reviewed (commented):
Thanks, left a high-level comment.

adisuissa commented:
/wait

adisuissa commented on source/common/filesystem/posix/filesystem_impl.cc:
```diff
@@ -339,9 +339,11 @@ bool InstanceImplPosix::illegalPath(const std::string& path) {
-  if (absl::StartsWith(canonical_path.return_value_, "/dev") ||
-      absl::StartsWith(canonical_path.return_value_, "/sys") ||
-      absl::StartsWith(canonical_path.return_value_, "/proc")) {
+  if (absl::StartsWith(canonical_path.return_value_, "/dev/") ||
```
Please update the comment above to reflect why the non-trivial comparison is used.

adisuissa reviewed (commented):
/assign-from @envoyproxy/senior-maintainers

adisuissa commented:
Assigning Harvey for a senior-maintainer pass.
/assign @htuch 

Waiting for comment update.
/wait

htuch reviewed (commented):
LGTM

zhangbo1882 commented:
@htuch , I removed the cases I added in last commit. Because in the CI test environment, there is no /sysroot directory, so the function canonicalPath can not return a valid path  so that /sysroot will always be considered as illegal, the test case will fail.
```
EXPECT_FALSE(file_system_.illegalPath("/sysroot/"));
```
In my local system, I have the /sysroot directory,  the test case will pass.
I do not have any good ideas to add the test cases in CI environment. 

I

htuch commented:
I think the whole test should move to a mock environment for filesystem impl, to avoid the CI machine dependency. This probably will save some grief in the future. We could leave one or two tests using the real `/dev` maybe, which should be standard, and mock the rest.

zhangbo1882 commented:
If we need to refactor the whole testing for filesystem impl, can we have another PR to do that?

htuch commented:
@zhangbo1882 that's fine if you're willing to do that.

zhangbo1882 commented:
I will try my best to do that. Do we need to create a issue for that ?

htuch reviewed (approved):
OK, LGTM providing there is some test followup.
